🐛 BUG: C3 crashes on git config if username & email are not set on host machine

### Which Cloudflare product(s) does this pertain to?

C3

### What version(s) of the tool(s) are you using?

2.1.0 [C3]

### What version of Node are you using?

v18.16.1

### What operating system are you using?

Linux (Fedora 38)

### Describe the Bug

I'm running C3 on a freshly installed machine where the global config for git has not been done, e.g. the two commands below have **not** been run:

```
  git config --global user.email "you@example.com"
  git config --global user.name "Your Name"
```

During the installation process, the CLI asks whether I would like to use git, I say yes and get the error message seen in the error below.

C3 should first check whether the required settings are in place before running into the error and forcing the user to re-do the entire process (e.g. selecting the stack again)


### Please provide a link to a minimal reproduction

_No response_

### Please provide any relevant error logs

├ Committing new files
│ Author identity unknown

*** Please tell me who you are.

Run

  git config --global user.email "you@example.com"
  git config --global user.name "Your Name"

to set your account's default identity.
Omit --global to set the identity only in this repository.

fatal: unable to auto-detect email address (got 'mxschumacher@fedora.(none)')

│
Error: Author identity unknown

*** Please tell me who you are.

Run

  git config --global user.email "you@example.com"
  git config --global user.name "Your Name"

to set your account's default identity.
Omit --global to set the identity only in this repository.

fatal: unable to auto-detect email address (got 'mxschumacher@fedora.(none)')

npm ERR! code 1
npm ERR! path /home/mxschumacher/code
npm ERR! command failed
npm ERR! command sh -c create-cloudflare

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/mxschumacher/.npm/_logs/2023-07-28T11_14_33_787Z-debug-0.log

--- a/.changeset/green-days-bow.md
+++ b/.changeset/green-days-bow.md
@@ -1,0 +1,5 @@
+---
+"create-cloudflare": patch
+---
+
+Don't prompt the user to use git if `user.name` and `user.email` haven't been configured

--- a/packages/create-cloudflare/src/__tests__/common.test.ts
+++ b/packages/create-cloudflare/src/__tests__/common.test.ts
@@ -1,0 +1,41 @@
+import * as command from "helpers/command";
+import { describe, expect, test, vi } from "vitest";
+import { isGitConfigured } from "../common";
+
+function promisify<T>(value: T) {
+	return new Promise<T>((res) => res(value));
+}
+
+describe("isGitConfigured", () => {
+	test("fully configured", async () => {
+		const spy = vi.spyOn(command, "runCommand");
+		spy.mockImplementation((cmd) =>
+			promisify(cmd.includes("email") ? "test@user.com" : "test user")
+		);
+		expect(await isGitConfigured()).toBe(true);
+	});
+
+	test("no name", async () => {
+		const spy = vi.spyOn(command, "runCommand");
+		spy.mockImplementation((cmd) =>
+			promisify(cmd.includes("email") ? "test@user.com" : "")
+		);
+		expect(await isGitConfigured()).toBe(false);
+	});
+
+	test("no email", async () => {
+		const spy = vi.spyOn(command, "runCommand");
+		spy.mockImplementation((cmd) =>
+			promisify(cmd.includes("name") ? "test user" : "")
+		);
+		expect(await isGitConfigured()).toBe(false);
+	});
+
+	test("runCommand fails", async () => {
+		const spy = vi.spyOn(command, "runCommand");
+		spy.mockImplementation(() => {
+			throw new Error("git not found");
+		});
+		expect(await isGitConfigured()).toBe(false);
+	});
+});

--- a/packages/create-cloudflare/src/common.ts
+++ b/packages/create-cloudflare/src/common.ts
@@ -242,12 +242,26 @@ export const printSummary = async (ctx: PagesGeneratorContext) => {
 
 export const offerGit = async (ctx: PagesGeneratorContext) => {
 	const gitInstalled = await isGitInstalled();
-
 	if (!gitInstalled) {
 		// haven't prompted yet, if provided as --git arg
 		if (ctx.args.git) {
 			updateStatus(
 				"Couldn't find `git` installed on your machine. Continuing without git."
+			);
+		}
+
+		// override true (--git flag) and undefined (not prompted yet) to false (don't use git)
+		ctx.args.git = false;
+
+		return; // bail early
+	}
+
+	const gitConfigured = await isGitConfigured();
+	if (!gitConfigured) {
+		// haven't prompted yet, if provided as --git arg
+		if (ctx.args.git) {
+			updateStatus(
+				"Must configure `user.name` and user.email` to use git. Continuing without git."
 			);
 		}
 
@@ -359,6 +373,26 @@ async function getGitVersion() {
  */
 async function isGitInstalled() {
 	return (await getGitVersion()) !== null;
+}
+
+async function isGitConfigured() {
+	try {
+		const userName = await runCommand("git config user.name", {
+			useSpinner: false,
+			silent: true,
+		});
+		if (!userName) return false;
+
+		const email = await runCommand("git config user.name", {
+			useSpinner: false,
+			silent: true,
+		});
+		if (!email) return false;
+
+		return true;
+	} catch {
+		return null;
+	}
 }
 
 /**

--- a/packages/create-cloudflare/src/common.ts
+++ b/packages/create-cloudflare/src/common.ts
@@ -371,11 +371,11 @@ async function getGitVersion() {
 /**
  * Check whether git is available on the user's machine.
  */
-async function isGitInstalled() {
+export async function isGitInstalled() {
 	return (await getGitVersion()) !== null;
 }
 
-async function isGitConfigured() {
+export async function isGitConfigured() {
 	try {
 		const userName = await runCommand("git config user.name", {
 			useSpinner: false,
@@ -383,7 +383,7 @@ async function isGitConfigured() {
 		});
 		if (!userName) return false;
 
-		const email = await runCommand("git config user.name", {
+		const email = await runCommand("git config user.email", {
 			useSpinner: false,
 			silent: true,
 		});

--- a/packages/create-cloudflare/src/common.ts
+++ b/packages/create-cloudflare/src/common.ts
@@ -391,7 +391,7 @@ async function isGitConfigured() {
 
 		return true;
 	} catch {
-		return null;
+		return false;
 	}
 }
 

